File: app/src/main/java/com/tkdev/weatherapp/presenter/forecast/ForecastPresenterImpl.java
package com.tkdev.weatherapp.presenter.forecast;

import com.tkdev.weatherapp.model.forecast_weather.ForecastRetrofit;
import com.tkdev.weatherapp.presenter.MainContract;
import com.tkdev.weatherapp.repository.WeatherRetrofitImpl;

import retrofit2.Response;

import static com.tkdev.weatherapp.utils.PreferencesVariables.current_city;
import static com.tkdev.weatherapp.utils.PreferencesVariables.last_dt;

public class ForecastPresenterImpl implements MainContract.Presenter, ForecastPresenter, MainContract.APIForecastListener {

    private MainContract.View view;
    private ForecastRetrofit forecasts;
    private MainContract.Model model;

    public ForecastPresenterImpl(MainContract.View view) {
        this.view = view;
        this.model = new WeatherRetrofitImpl();
        this.forecasts = new ForecastRetrofit();
    }

    public ForecastRetrofit getForecasts() {
        return forecasts;
    }

    @Override
    public void onSuccessResponse(Response<ForecastRetrofit> response) {
        forecasts = response.body();
        view.update();
    }



    @Override
    public void onFailureResponse(String responseMessage) {
//        view.onFailUpdate(responseMessage);
    }


    @Override
    public void onDestroy() {
        view = null;
    }

    @Override
    public void onRequestWeather(String city) {
        if (current_city.equals("")) {
            current_city = "london";
            model.getWeather(this, current_city);
        } else if (
                last_dt <= System.currentTimeMillis() / 1000 - 600
                        ||
                        (!(current_city.equals(city))))
        {
            model.getWeather(this, city);
        } else {
            view.cancelUpdate();
        }
    }
}

File: app/src/main/java/com/tkdev/weatherapp/presenter/forecast/ForecastPresenterImpl.kt
package com.tkdev.weatherapp.presenter.forecast

import com.tkdev.weatherapp.model.forecast_weather.ForecastRetrofit
import com.tkdev.weatherapp.presenter.MainContract

import com.tkdev.weatherapp.repository.WeatherRetrofitImpl
import com.tkdev.weatherapp.utils.PreferencesVariables
import retrofit2.Response

class ForecastPresenterImpl(private var view: MainContract.View) : MainContract.Presenter, ForecastPresenter, MainContract.APIForecastListener {

    var forecasts: ForecastRetrofit
    private val model: MainContract.Model

    init {
        model = WeatherRetrofitImpl()
        forecasts = ForecastRetrofit()
    }

    override fun onSuccessResponse(response: Response<ForecastRetrofit>) {
        forecasts = response.body()!!
        view.update()
    }

    override fun onFailureResponse(responseMessage: String) {
//        view.onFailUpdate(responseMessage);
    }

    override fun onDestroy() {

    }

    override fun onRequestWeather(city: String) {
        if (PreferencesVariables.current_city == "") {
            PreferencesVariables.current_city = "london"
            model.getWeather(this, PreferencesVariables.current_city)
        } else if (PreferencesVariables.last_dt <= System.currentTimeMillis() / 1000 - 600
                ||
                PreferencesVariables.current_city != city) {
            model.getWeather(this, city)
        } else {
            view.cancelUpdate()
        }
    }
}